beluga_amcl/src/ndt_amcl_node_3d.cpp

```diff
@@ -0,0 +1,503 @@
+        *laser_scan_sub_, *tf_buffer_, get_parameter("odom_frame_id").as_string(), 10, get_node_logging_interface(),
+        get_node_clock_interface(), tf2::durationFromSec(get_parameter("transform_tolerance").as_double()));
+
+    laser_scan_connection_ =
+        laser_scan_filter_->registerCallback(std::bind(&NdtAmclNode3D::laser_callback, this, std::placeholders::_1));
+    RCLCPP_INFO(get_logger(), "Subscribed to scan_topic: %s", laser_scan_sub_->getTopic().c_str());
+
+    const auto initial_estimate = get_initial_estimate();
+    if (initial_estimate.has_value()) {
+      last_known_estimate_ = initial_estimate;
+      last_known_odom_transform_in_map_.reset();
+      initialize_from_estimate(initial_estimate.value());
+    }
+  }
+}
+
+void NdtAmclNode3D::do_deactivate(const rclcpp_lifecycle::State&) {
+  RCLCPP_INFO(get_logger(), "Cleaning up");
+  particle_cloud_pub_.reset();
+  laser_scan_connection_.disconnect();
+  laser_scan_filter_.reset();
+  laser_scan_sub_.reset();
+}
+
+void NdtAmclNode3D::do_cleanup(const rclcpp_lifecycle::State&) {
+  RCLCPP_INFO(get_logger(), "Cleaning up");
+  particle_cloud_pub_.reset();
+  pose_pub_.reset();
+  particle_filter_.reset();
+  enable_tf_broadcast_ = false;
+}
+
+auto NdtAmclNode3D::get_initial_estimate() const -> std::optional<std::pair<Sophus::SE3d, Sophus::Matrix6d>> {
+  if (!get_parameter("set_initial_pose").as_bool()) {
+    return std::nullopt;
+  }
+
+  const auto pose = Sophus::SE3d{
+      Sophus::SO3d::rotZ(get_parameter("initial_pose.yaw").as_double()) *
+          Sophus::SO3d::rotY(get_parameter("initial_pose.pitch").as_double()) *
+          Sophus::SO3d::rotX(get_parameter("initial_pose.roll").as_double()),
+      Eigen::Vector3d{
+          get_parameter("initial_pose.x").as_double(),
+          get_parameter("initial_pose.y").as_double(),
+          get_parameter("initial_pose.z").as_double(),
+      }};
+
+  Sophus::Matrix6d covariance = Sophus::Matrix6d::Zero();
+
+  // TODO(serraramiro1): We are ignoring correlations here.
+  // Is it realistic for a user to initialize a filter with correlations?
+  covariance.coeffRef(0, 0) = get_parameter("initial_pose.covariance_x").as_double();
+  covariance.coeffRef(1, 1) = get_parameter("initial_pose.covariance_y").as_double();
+  covariance.coeffRef(2, 2) = get_parameter("initial_pose.covariance_z").as_double();
+  covariance.coeffRef(3, 3) = get_parameter("initial_pose.covariance_roll").as_double();
+  covariance.coeffRef(4, 4) = get_parameter("initial_pose.covariance_pitch").as_double();
+  covariance.coeffRef(5, 5) = get_parameter("initial_pose.covariance_yaw").as_double();
+
+  return std::make_pair(pose, covariance);
+}
+
+auto NdtAmclNode3D::get_motion_model() const -> MotionModelVariant {
+  const auto name = get_parameter("robot_model_type").as_string();
+  if (name == kDifferentialModelName || name == kNav2DifferentialModelName) {
```

Comment:
@serraramiro1 meta: I would just drop `kNav2DifferentialModelName` compatibility here. This is not good ol' AMCL.